### ISSUE TITLE
Decide on many-to-many representation in schema.graphql

### ISSUE THREAD

**Issue by eldargab** [subsquid/squid-sdk#3]
Many-to-many relations in [the graph](https://thegraph.com/docs/en/developer/create-subgraph-hosted/#many-to-many-relationships) and hydra v4 supposed to be defined as follows:

```graphql
type Organization @entity {
  id: ID!
  name: String!
  members: [User!]! @derivedFrom(field: "organization")
}

type User @entity {
  id: ID!
  name: String!
  organizations: [Organization!]! 
}
```

However, this seems to be very problematic.

First, `organizations: [Organization!]!` property can be easily confused with a real array of organization ids.
Such implementation of many-to-many relation makes perfect sense and very useful for some cases.
Although, we don't plan to support this in near future, it would be nice to reserve the possibility.

Second, the above scheme is needlessly obscure. It hides the underlying join table, creating possibility for name conflict,
also join table generation algorithm is not clear.

Third, it makes impossible to add additional fields to the join table.

In summary:

Pros:

* compatibility with the graph and hydra v4

Cons:

* breaks the nice rule when all virtual lookup fields have `@deriveFrom`
* obscure
* makes arrays of ids ugly to express
* makes addition of extra fields to the join table ugly to express
* makes impossible to have multiple kinds of relations involving same entities.

## Suggestion

To define many-to-many relation one have to define join table explicitly:

```graphql
type UserOrganization @entity {
  user: User!
  organization: Organization!
}

type Organization @entity {
  id: ID!
  name: String!
  members: [User!]! @derivedFrom(field: "organization", joinTable: "UserOrganization")
}

type User @entity {
  id: ID!
  name: String!
  organizations: [Organization!]! @derivedFrom(field: "member", joinTable: "UserOrganization")
}
```

We can decide on rules for primary keys of join tables and other details after we conclude that this is a general way to go.

**Comment by ScarfaceIII** [comment]
Here's my two cents:

- I think that the distinction between a many-to-many that has objects relations can be distinguished from a real array of ids, by making it explicit in the field name. In the provided example, it can be `organization_ids: [ID!]!`, what do you think? (it might be ugly, as you said 😆 )
- I understand it might "seem" obscure, but many high level frameworks handle underlying join tables transparently, I can think of Django, which is a Python framework I have used, for example.
- As for additional fields in the join table I don't have a counter.
- This being said, the explicit join table does not seem like an inconvenient proposal, the provided example code seems straightforward, concise and consistent

In conclusion, I would go for the implicit join table for ease of use and simplicity approach where the dev does not need to know what's under the hood.
The explicit join table allows for more customization, makes devs more aware of what's going on and seems to be more performant, as well, according to the Graph doc (will it be the same for us? I don't see why not...), at least. So I have a slight preference for the explicit version.

**Comment by dzhelezov** [comment]
@eldargab Have we decided on that ?

**Comment by RaekwonIII** [comment]
Follow up, looking at things for the reference documentation, I came across this
https://docs.fauna.com/fauna/current/api/graphql/relationships#one2many
They use a `@relation` directive to explicit that they *don't* want the field to be an array of IDs. We could either do that, or make this the default behavior and have a directive for when someone explicitly wants an array of IDs.
